File: sdks/typescript/trace-harness/src/analyze/http.ts
import { httpRequests, type HttpRequest } from "../kinds/http";
import type { Finding } from "../model/node";
import type { Detector } from "./registry";

const SLOW_HTTP_MS = 200;
const MAX_FINDINGS = 10;
const duration = (request: HttpRequest) => Math.max(...request.spans.map((span) => span.dur_ms));
const label = (request: HttpRequest) => `${request.api[0]} ${request.api[1]}${request.api[2]}`;
const compareId = (a: string, b: string) => a < b ? -1 : a > b ? 1 : 0;
const wall = (run: HttpRequest[]) => run.at(-1)!.span.end_ms - run[0]!.span.start_ms;

function serialRuns(requests: HttpRequest[]): HttpRequest[][] {
  const groups = new Map<string, HttpRequest[]>();
  for (const request of requests) {
    if (!request.span.parent_span_id) continue;
    const key = JSON.stringify([request.span.service, request.span.parent_span_id]);
    const group = groups.get(key) ?? [];
    group.push(request);
    groups.set(key, group);
  }
  const runs: HttpRequest[][] = [];
  for (const siblings of groups.values()) {
    let run: HttpRequest[] = [];
    let busyUntil = -Infinity;
    for (const request of siblings.sort((a, b) => a.span.start_ms - b.span.start_ms || compareId(a.span.span_id, b.span.span_id))) {
      const serial = request.span.start_ms >= busyUntil;
      if (request.ordinary && serial && run.length && request.api.every((value, index) => value === run.at(-1)!.api[index])) {
        run.push(request);
      } else {
        if (run.length >= 2) runs.push(run);
        run = request.ordinary && serial ? [request] : [];
      }
      busyUntil = Math.max(busyUntil, request.span.end_ms);
    }
    if (run.length >= 2) runs.push(run);
  }
  return runs;
}

/** Run once per trace; use caller spans for sequence timing to avoid cross-host clock skew. */
export const httpRequestPatterns: Detector = (node, context) => {
  const anchor = context.nodes.reduce((earliest, candidate) =>
    !earliest || candidate.start_ms < earliest.start_ms
      || (candidate.start_ms === earliest.start_ms && candidate.node_id < earliest.node_id)
      ? candidate : earliest, context.nodes[0]);
  if (!anchor || node.node_id !== anchor.node_id) return [];
  const requests = httpRequests(context);
  const view = context.view();
  const findings: Finding[] = [];
  const slow = requests.filter((request) => request.ordinary && duration(request) > SLOW_HTTP_MS)
    .sort((a, b) => duration(b) - duration(a) || compareId(a.span.span_id, b.span.span_id));
  for (const request of slow.slice(0, MAX_FINDINGS)) {
    const owner = request.spans.map((span) => view.by_span.get(span.span_id)).find(Boolean) ?? node;
    findings.push({
      ref: owner.node_id,
      source: "http_slow_request",
      severity: "warn",
      symptoms: ["慢"],
      note: `普通 HTTP 请求 ${label(request)} 耗时 ${duration(request).toFixed(1)}ms > ${SLOW_HTTP_MS}ms；检查 span ${request.span.span_id} 的下游调用与请求内空档`,
      data: { method: request.api[0], target: request.api[1], route: request.api[2], duration_ms: duration(request), threshold_ms: SLOW_HTTP_MS, span_ids: request.spans.map((span) => span.span_id) },
    });
  }
  const runs = serialRuns(requests).sort((a, b) => wall(b) - wall(a));
  for (const run of runs.slice(0, MAX_FINDINGS)) {
    const first = run[0]!;
    const total = run.reduce((sum, request) => sum + request.span.dur_ms, 0);
    const gap = run.slice(1).reduce((sum, request, index) => sum + request.span.start_ms - run[index]!.span.end_ms, 0);
    const owner = view.by_span.get(first.span.parent_span_id!) ?? view.by_span.get(first.span.span_id) ?? node;
    findings.push({
      ref: owner.node_id,
      source: "http_serial_same_api",
      severity: "warn",
      symptoms: ["慢"],
      note: `${first.span.service ?? "?"} 连续串行调用 ${label(first)} ${run.length} 次，wall-clock ${wall(run).toFixed(1)}ms（请求累计 ${total.toFixed(1)}ms，间隔 ${gap.toFixed(1)}ms）；检查批量接口、请求内结果复用与 client 生命周期；同一 API 不代表参数相同；检查父 span ${first.span.parent_span_id} 的调用上下文`,
      data: { method: first.api[0], target: first.api[1], route: first.api[2], count: run.length, wall_ms: wall(run), gap_ms: gap, http_total_ms: total, parent_span_id: first.span.parent_span_id, span_ids: run.map((request) => request.span.span_id) },
    });
  }
  for (const [source, total] of [["http_slow_request", slow.length], ["http_serial_same_api", runs.length]] as const) {
    if (total > MAX_FINDINGS) findings.find((finding) => finding.source === source)!.note += `（共 ${total} 条，仅展示耗时最高的 ${MAX_FINDINGS} 条）`;
  }
  return findings;
};

File: sdks/python/trace_harness/analyze/diagnose/http.py
"""通用 HTTP 慢请求和连续串行同 API 调用判读。"""

from collections import defaultdict

from trace_harness.analyze.context import AnalysisContext
from trace_harness.kinds.http import HttpRequest, http_requests
from trace_harness.model.node import Finding, Node

_SLOW_HTTP_MS = 200
_MAX_FINDINGS = 10


def _serial_runs(requests: list[HttpRequest]) -> list[list[HttpRequest]]:
    groups: dict[tuple[str, str], list[HttpRequest]] = defaultdict(list)
    for request in requests:
        if request.span.parent_span_id:
            groups[(request.span.service, request.span.parent_span_id)].append(request)
    runs = []
    for siblings in groups.values():
        run = []
        busy_until = float("-inf")
        for request in sorted(siblings, key=lambda item: (item.start_ms, item.span.span_id)):
            serial = request.start_ms >= busy_until
            if request.ordinary and serial and run and request.api == run[-1].api:
                run.append(request)
            else:
                if len(run) >= 2:
                    runs.append(run)
                run = [request] if request.ordinary and serial else []
            # 并发区间里的后续短调用也不能被误认成一段串行序列。
            busy_until = max(busy_until, request.end_ms)
        if len(run) >= 2:
            runs.append(run)
    return runs


async def http_request_patterns(node: Node, analysis: AnalysisContext) -> list[Finding]:
    ctx = analysis.trace
    """同一 HTTP 调用 client/server 去重；两个独立 warn 各保留耗时最高的 10 条。"""
    if (
        not ctx.nodes
        or node.node_id != min(ctx.nodes, key=lambda item: (item.start_ms, item.node_id)).node_id
    ):
        return []
    await analysis.fact(node, "http_evidence")
    requests = http_requests(ctx)
    view = ctx.view()
    findings = []
    slow = sorted(
        (item for item in requests if item.ordinary and item.duration_ms > _SLOW_HTTP_MS),
        key=lambda item: (-item.duration_ms, item.span.span_id),
    )
    for request in slow[: analysis.finding_limit]:
        owner = next(
            (view.by_span[item.span_id] for item in request.spans if item.span_id in view.by_span),
            node,
        )
        findings.append(
            Finding(
                owner.node_id,
                "http_slow_request",
                "warn",
                symptoms=("慢",),
                note=(
                    f"普通 HTTP 请求 {request.label} "
                    f"耗时 {request.duration_ms:.1f}ms > {_SLOW_HTTP_MS}ms；"
                    f"检查 span {request.span.span_id} 的下游调用与请求内空档"
                ),
                data={
                    "method": request.api[0],
                    "target": request.api[1],
                    "route": request.api[2],
                    "duration_ms": request.duration_ms,
                    "threshold_ms": _SLOW_HTTP_MS,
                    "span_ids": [span.span_id for span in request.spans],
                },
            )
        )
    runs = sorted(_serial_runs(requests), key=lambda run: -(run[-1].end_ms - run[0].start_ms))
    for run in runs[: analysis.finding_limit]:
        first, last = run[0], run[-1]
        wall_ms = last.end_ms - first.start_ms
        gap_ms = sum(
            right.start_ms - left.end_ms for left, right in zip(run, run[1:], strict=False)
        )
        total_ms = sum(item.span.dur_ms for item in run)
        owner = (
            view.by_span.get(first.span.parent_span_id)
            or view.by_span.get(first.span.span_id)
            or node
        )
        findings.append(
            Finding(
                owner.node_id,
                "http_serial_same_api",
                "warn",
                symptoms=("慢",),
                note=(
                    f"{first.span.service or '?'} 连续串行调用 {first.label} {len(run)} 次，"
                    f"wall-clock {wall_ms:.1f}ms（请求累计 {total_ms:.1f}ms，"
                    f"间隔 {gap_ms:.1f}ms）；"
                    "检查批量接口、请求内结果复用与 client 生命周期；同一 API 不代表参数相同；"
                    f"检查父 span {first.span.parent_span_id} 的调用上下文"
                ),
                data={
                    "method": first.api[0],
                    "target": first.api[1],
                    "route": first.api[2],
                    "count": len(run),
                    "wall_ms": wall_ms,
                    "gap_ms": gap_ms,
                    "http_total_ms": total_ms,
                    "parent_span_id": first.span.parent_span_id,
                    "span_ids": [item.span.span_id for item in run],
                },
            )
        )
    for source, total in (
        ("http_slow_request", len(slow)),
        ("http_serial_same_api", len(runs)),
    ):
        if total > _MAX_FINDINGS:
            first = next(item for item in findings if item.source == source)
            first.note += f"（共 {total} 条，仅展示耗时最高的 {_MAX_FINDINGS} 条）"
    return findings
